ospackage/recipes/artifacts.go

```diff
@@ -0,0 +1,186 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package recipes
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	osconfigpb "github.com/GoogleCloudPlatform/osconfig/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/v1alpha2"
+)
+
+var (
+	testStorageClient *storage.Client
+	testHTTPClient    *http.Client
+)
+
+func newStorageClient(ctx context.Context) (*storage.Client, error) {
+	if testStorageClient != nil {
+		return testStorageClient, nil
+	}
+	return storage.NewClient(ctx)
+}
+
+func newHTTPClient() *http.Client {
+	if testHTTPClient != nil {
+		return testHTTPClient
+	}
+	return &http.Client{}
+}
+
+// FetchArtifacts takes in a slice of artifacs and dowloads them into the specified directory,
+// Returns a map of artifact names to their new locations on the local disk.
+func FetchArtifacts(ctx context.Context, artifacts []*osconfigpb.SoftwareRecipe_Artifact, directory string) (map[string]string, error) {
+	localNames := make(map[string]string)
+
+	for _, a := range artifacts {
+		path, err := fetchArtifact(ctx, a, directory)
+		if err != nil {
+			return nil, err
+		}
+		localNames[a.Id] = path
+	}
+
+	return localNames, nil
+}
+
+func fetchArtifact(ctx context.Context, artifact *osconfigpb.SoftwareRecipe_Artifact, directory string) (string, error) {
+	localPath := path.Join(directory, artifact.Id)
+	uri, err := url.Parse(artifact.Uri)
+	if err != nil {
+		return "", fmt.Errorf("Could not parse url %q for artifact %q", artifact.Uri, artifact.Id)
+	}
+
+	var reader io.ReadCloser
+	switch strings.ToLower(uri.Scheme) {
+	case "gcs":
+		reader, err = fetchWithGCS(ctx, uri.Host, uri.Path, uri.Fragment)
+		defer reader.Close()
+		if err != nil {
+			return "", fmt.Errorf("error fetching artifact %q from GCS: %v", artifact.Id, err)
+		}
+	case "https", "http":
+		reader, err = fetchWithHTTP(ctx, artifact.Uri)
+		defer reader.Close()
+		if err != nil {
+			return "", fmt.Errorf("error fetching artifact %q with http or https: %v", artifact.Id, err)
+		}
+	default:
+		return "", fmt.Errorf("artifact %q has unsupported protocol %s", artifact.Id, uri.Scheme)
+	}
+
+	err = downloadStream(reader, artifact.Checksum, localPath)
```

Comment:
For GCS checksum and generation are redundant, maybe this is ok to just be solved with documentation in the API by providing checksum for GCS objects should be discouraged

Reply:
yeah, my thought was that the api will just do as it's told if it gets both a generation and a checksum.